File: e3_django/compute/objects/SensitivitySummary.py
class SensitivitySummary():
    """
    Represents a sensitivity summary object.
    """
    def __init__(self, globalVarBool, bcnObj, varName, diffType, diffVal, diffSign, measure_summaries):
        self.globalVarBool = globalVarBool
        self.bcnObj = bcnObj
        self.varName = varName
        self.diffType = diffType
        self.diffVal = diffVal
        self.diffSign = diffSign
        ## self.altOutput = {measure_summaries.altID: measure_summaries for measure_summaries in measure_summaries}
        self.totalBenefits = {measure_summary.altID: measure_summary.totalBenefits for measure_summary in measure_summaries}
        self.totalCosts = {measure_summary.altID: measure_summary.totalCosts for measure_summary in measure_summaries}
        self.totalCostsInv = {measure_summary.altID: measure_summary.totalCostsInv for measure_summary in measure_summaries}
        self.totalCostsNonInv = {measure_summary.altID: measure_summary.totalCostsNonInv for measure_summary in measure_summaries}
        self.totSubtypeFlows = {measure_summary.altID: measure_summary.totSubtypeFlows for measure_summary in measure_summaries}
        self.totTagFlows = {measure_summary.altID: measure_summary.totTagFlows for measure_summary in measure_summaries}
        self.netBenefits = {measure_summary.altID: measure_summary.netBenefits for measure_summary in measure_summaries}
        self.netSavings = {measure_summary.altID: measure_summary.netSavings for measure_summary in measure_summaries}
        self.SIR = {measure_summary.altID: measure_summary.SIR for measure_summary in measure_summaries}
        self.IRR = {measure_summary.altID: measure_summary.IRR for measure_summary in measure_summaries}
        self.AIRR = {measure_summary.altID: measure_summary.AIRR for measure_summary in measure_summaries}
        self.DPP = {measure_summary.altID: measure_summary.DPP for measure_summary in measure_summaries}
        self.SPP = {measure_summary.altID: measure_summary.SPP for measure_summary in measure_summaries}
        self.BCR = {measure_summary.altID: measure_summary.BCR for measure_summary in measure_summaries}
        self.quantSum = {measure_summary.altID: measure_summary.quantSum for measure_summary in measure_summaries}
        self.quantUnits = {measure_summary.altID: measure_summary.quantUnits for measure_summary in measure_summaries}
        self.MARR = {measure_summary.altID: measure_summary.MARR for measure_summary in measure_summaries}
        self.deltaQuant = {measure_summary.altID: measure_summary.deltaQuant for measure_summary in measure_summaries}
        self.nsDeltaQuant = {measure_summary.altID: measure_summary.nsDeltaQuant for measure_summary in measure_summaries}
        self.nsPercQuant = {measure_summary.altID: measure_summary.nsPercQuant for measure_summary in measure_summaries}
        self.nsElasticityQuant = {measure_summary.altID: measure_summary.nsElasticityQuant for measure_summary in measure_summaries}

    ## For testing, has no use in final code
    def __str__(self):
        print("globalVarBool:", self.globalVarBool)
        print("bcnObj:", self.bcnObj)
        print("varName:", self.varName)
        print("diffType:", self.diffType)
        print("diffVal:", self.diffVal)
        print("diffSign:", self.diffSign)
        print("totalBenefits:", self.totalBenefits)
        print("totalCosts:", self.totalCosts)
        print("totalCostsInv:", self.totalCostsInv)
        print("totalCostsNonInv:", self.totalCostsNonInv)
        print("totTagFlows:", self.totTagFlows)
        print("netBenefits:", self.netBenefits)
        print("netSavings:", self.netSavings)
        print("SIR:", self.SIR)
        print("IRR:", self.IRR)
        print("AIRR:", self.AIRR)
        print("SPP:", self.SPP)
        print("DPP:", self.DPP)
        print("BCR:", self.BCR)
        print("quantSum:", self.quantSum)
        print("quantUnits", self.quantUnits)
        print("MARR:", self.MARR)
        print("deltaQuant:", self.deltaQuant)
        print("nsPercQuant:", self.nsPercQuant)
        print("nsDeltaQuant:", self.nsDeltaQuant)
        print("nsElasticityQuant:", self.nsElasticityQuant)
        return "--------------------End of Object--------------------"

File: src/main/java/gov/nist/e3/objects/output/SensitivitySummary.java
package gov.nist.e3.objects.output;

import gov.nist.e3.util.Util;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public record SensitivitySummary(
        int id,
        int altId,
        List<Double> totalBenefits,
        List<Double> totalCosts,
        List<Double> totalCostsInvest,
        List<Double> totalCostNonInvest,
        Map<String, List<Double>> totalTagFlows,
        List<Double> netBenefits,
        List<Double> netSavings,
        List<Double> sir,
        List<Double> irr,
        List<Double> airr,
        List<Double> dpp,
        List<Double> spp,
        List<Double> bcr,
        Map<String, List<Double>> quantitySum,
        Map<String, List<String>> quantityUnits,
        List<Double> marr,
        Map<String, List<Double>> deltaQuantity,
        Map<String, List<Double>> nsPercentQuantity,
        Map<String, List<Double>> nsDeltaQuantity,
        Map<String, List<Double>> nsElasticityQuantity
) {
    public static SensitivitySummary of(int id, MeasureSummary summary) {
        return new SensitivitySummary(
                id,
                summary.altId(),
                Util.createList(summary.totalBenefits()),
                Util.createList(summary.totalCosts()),
                Util.createList(summary.totalCostsInvest()),
                Util.createList(summary.totalCostNonInvest()),
                toMapList(summary.totalTagFlows()),
                Util.createList(summary.netBenefits()),
                Util.createList(summary.netSavings()),
                Util.createList(summary.sir()),
                Util.createList(summary.irr()),
                Util.createList(summary.airr()),
                Util.createList(summary.dpp()),
                Util.createList(summary.spp()),
                Util.createList(summary.bcr()),
                toMapList(summary.quantitySum()),
                toMapList(summary.quantityUnits()),
                Util.createList(summary.marr()),
                toMapList(summary.deltaQuantity()),
                toMapList(summary.nsPercentQuantity()),
                toMapList(summary.nsDeltaQuantity()),
                toMapList(summary.nsElasticityQuantity())
        );
    }

    public void addMeasureSummary(MeasureSummary summary) {
        if (summary.altId() != this.altId())
            throw new IllegalArgumentException("Cannot add measure summary to sensitivity summary with different alt IDs");

        this.totalBenefits.add(summary.totalBenefits());
        this.totalCosts.add(summary.totalCosts());
        this.totalCostsInvest.add(summary.totalCostsInvest());
        this.totalCostNonInvest.add(summary.totalCostNonInvest());

        addEntries(summary.totalTagFlows(), this.totalTagFlows());

        this.netBenefits.add(summary.netBenefits());
        this.netSavings.add(summary.netSavings());
        this.sir.add(summary.sir());
        this.irr.add(summary.irr());
        this.airr.add(summary.airr());
        this.dpp.add(summary.dpp());
        this.spp.add(summary.spp());
        this.bcr.add(summary.bcr());

        addEntries(summary.quantitySum(), this.quantitySum());
        addEntries(summary.quantityUnits(), this.quantityUnits());

        this.marr.add(summary.marr());

        addEntries(summary.deltaQuantity(), this.deltaQuantity());
        addEntries(summary.nsPercentQuantity(), this.nsPercentQuantity());
        addEntries(summary.nsDeltaQuantity(), this.nsDeltaQuantity());
        addEntries(summary.nsElasticityQuantity(), this.nsElasticityQuantity());
    }

    private static <A, B> void addEntries(Map<A, B> toAdd, Map<A, List<B>> to) {
        if (toAdd == null)
            return;

        for (var entry : toAdd.entrySet()) {
            to.get(entry.getKey()).add(entry.getValue());
        }
    }

    private static <A, B> Map<A, List<B>> toMapList(Map<A, B> original) {
        var result = new HashMap<A, List<B>>();

        if (original != null) {
            for (var entry : original.entrySet()) {
                result.put(entry.getKey(), Util.createList(entry.getValue()));
            }
        }

        return result;
    }
}
